osquery/tables/system/windows/ads.cpp

```diff
@@ -0,0 +1,193 @@
+  Row r;
+  r["path"] = path;
+  r["directory"] = boost::filesystem::path(path).parent_path().string();
+  r["key"] = key;
+  if (isPrintable(value)) {
+    r["value"] = value;
+    r["base64"] = INTEGER(0);
+  } else {
+    r["value"] = base64::encode(value);
+    r["base64"] = INTEGER(1);
+  }
+  results.push_back(r);
+}
+
+void parseZoneIdentifier(QueryData& results,
+                         const std::string& path,
+                         const std::string& streamData) {
+  auto lines = split(streamData, "\n");
+  for (const auto& line : lines) {
+    auto key_len = line.find_first_of("=");
+    if (key_len == std::string::npos) {
+      continue;
+    }
+
+    setRow(results,
+           path,
+           line.substr(0, key_len),
+           line.substr(key_len + 1, line.size()));
+  }
+}
+
+// Process a file and extract all stream names and data.
+void enumerateStreams(QueryData& results, const std::string& path) {
+  WIN32_FIND_STREAM_DATA findStreamData;
+  HANDLE hFind = FindFirstStreamW(stringToWstring(path).c_str(),
+                                  FindStreamInfoStandard,
+                                  &findStreamData,
+                                  0);
+
+  auto fd_guard = scope_guard::create([&] { FindClose(hFind); });
+
+  if (hFind != INVALID_HANDLE_VALUE) {
+    do {
+      std::string stream(wstringToString(findStreamData.cStreamName));
+
+      // Split the stream string into a name and a type, format is
+      // ":streamname:$streamtype"
+      auto streamFullName = split(stream, ":");
+
+      if (streamFullName.size() != 2) {
+        LOG(WARNING) << "Invalid stream name found: '" << stream
+                     << "'. Skipping this entry";
+        continue;
+      }
+      std::string streamName = streamFullName[0];
+
+      // Skip unnamed stream since it represents the file content
+      if (streamName == "") {
+        continue;
+      }
+
+      std::string path_copy = path;
+      // Remove any potential trailing / from path string
+      if (boost::algorithm::ends_with(path_copy, "\\")) {
+        path_copy.pop_back();
+      }
+      std::stringstream streamPath;
+      streamPath << path_copy << ":" << streamName;
+      std::string streamData;
+
+      if (!readFile(streamPath.str(), streamData).ok()) {
+        LOG(INFO) << "Couldn't read stream data: " << streamPath.str();
+        continue;
+      }
+
+      if (streamName == kZoneIdentifierKey) {
+        parseZoneIdentifier(results, path, streamData);
+      } else {
+        // Remove trailing newlines
+        boost::trim_right(streamData);
+        setRow(results, path, streamName, streamData);
+      }
+    } while (FindNextStreamW(hFind, &findStreamData));
+  } else {
+    auto error_code = GetLastError();
+    if (error_code != ERROR_HANDLE_EOF) {
+      LOG(INFO) << "Error occurred while searching for streams in " << path
+                << ". Error code: " << error_code;
+    }
+  }
+}
+
+QueryData genAds(QueryContext& context) {

```

Comment:
I'm a little unsure of how the `path` and `directory` expansion logic work. 

I _think_ the intent is when one can query a directory, and get back all the files inside. (Similar to the `file` table). Or one can query a specific path. (And both probably support `LIKE`). This makes sense to me.

But the implementation enumerates the the `path` first, and then the `directory`. And I think that will result in equal work being done. Sometimes? I guess it's confusing.

If both are part of the query predicate, sqlite will filter to only return rows that match _both_. Eg `select * from ads where directory = '/tmp/' AND path = '/var/tmp/foo'` would result first in enumerating `/var/tmp/foo`, then all the files in `/tmp`, and would ultimately return nothing, because sqlite filtered them.

Though, to correct myself, that's not at all true there's an `OR` in there. 

I guess I'd suggest a reasonable pattern is to generate the list of things to enumerate, and then find the union of them. This would, at least, prevent duplicate enumeration.

Reply:
The logic is a copy paste from the [extended_attributes table](https://github.com/osquery/osquery/blob/master/osquery/tables/system/darwin/extended_attributes.cpp#L193-L256) since it was from where I based by work form. I agree is a bit confusing 😅 

I can take a look into making it more efficient.